kpr: Fix unconnected UDP backend selection on service endpoint updates

aditighag opened the pull request:
bpf: Fix backend selection for unconnected UDP service endpoint updates

In case of socket-lb for unconnected UDP, we select a service backend
based on the service backends count. So in case of updates, we may end
up selecting a wrong backend for existing UDP connections.

Fix:

When session affinity is disabled, we reuse the session affinity map
to store mapping for a UDP client to its selected backend using socket
cookie.
Why not enable session affinity?
Session affinity is a stronger affinity in the sense that all subsequent
connections from a client will go to the same selected backend. This can
lead to uneven load-balancing.

See commit description for more details.

Upgrade/Downgrade considerations:

Introduce a new flag in affinity map key using reserved bits to mark an
entry in order to indicate if its storing netns or socket cookie. In case of
downgrade, such entries won't be read as the flag wouldn't exist.


Fixes: https://github.com/cilium/cilium/issues/11301

```release-note
Fix connectivity issues for UDP connections on service endpoints' updates.
Disabling `EnableSessionConfig` won't disable session affinity feature.
```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on bpf/bpf_sock.c:
```diff
@@ -389,6 +390,17 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
 				backend_id = 0;
 		}
+	} else if (udp_only) {
+		udp_sock_affinity = true;
```
Doesn't `udp_only` imply `udp_sock_affinity = true`?

borkmann replied:
Also can we fold this whole path into the above? It's pretty much the same logic, except for the id population and `lb4_affinity_backend_id_by_netns` vs `lb4_affinity_backend_id_by_sock` retrieval. Maybe this could even be refactored as `lb4_affinity_backend_id(svc, &id, {SOCK,NETNS})`.

borkmann replied:
Also, IPv6 bits are missing.

aditighag replied:
Re IPv6 changes : I first wanted to check if I don't run into any verifier complexity issues on ipv4.

borkmann commented on bpf/lib/common.h:
```diff
@@ -827,13 +827,15 @@ struct ipv4_revnat_entry {
 	__u16 rev_nat_id;
 	__u8 netns_cookie:1,
-	     reserved:7;
+		 sock_cookie:1,
```
nit: whitespace. But otherwise this and lb4_affinity_client_id looks good.

borkmann commented on bpf/lib/lb.h:
```diff
@@ -1309,6 +1309,56 @@ lb4_update_affinity_by_netns(const struct lb4_service *svc __maybe_unused,
+	struct lb4_affinity_key key = {
+			.rev_nat_id   = svc->rev_nat_index,
+			.sock_cookie  = true,
+			.client_id    = *id,
```
nit: whitespace. `sock_cookie  = 1` given it's bit and not bool.

borkmann reviewed (changes requested):
Good starting point, left some comments for todo. Overall we should get rid of `ENABLE_SESSION_AFFINITY` and always build it in.

aditighag commented on bpf/bpf_sock.c:
```diff
@@ -367,33 +368,20 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
+	 * socket, we continue to select the same backend even when
+	 * the service backends are updated.
+	 */
+	if (udp_only && !svc_has_affinity) {
```
@borkmann I was in two minds whether to wrap this under `BPF_HAVE_SOCK_COOKIE` check as getting rid of this `if` block could help with verifier complexity on older kernels.

aditighag committed:
test: Validate UDP connectivity on service endpoints updates

The test validates that (unconnected) UDP connectivity
is uninterrupted when service endpoints are updated.
Specifically, the test checks if a client pod reported
a mismatch in the server reply after service endpoints
are updated. Also, checks if the client pod's connectivity was
disrupted, and as a result, it was restarted during
the test. Both the cases are indication that client's
subsequent messages were sent to a different UDP endpoint.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
config: Enable affinity related BPF code by default

The `ENABLE_SESSION_AFFINITY` macro in the datapath was previously
tied to the enable-session-affinity config option. This is unnecessary
as the related code will be executed if a user has enabled affinity for
a service. Additionally, this is also a preparatory commit as the affinity
maps are reused in the socket-lb to store UDP client socket to backend
mapping in order to keep connections uninterrupted when service backends change.

Hence, we replace agent config check to enable the macro to enable
the {socket, session} affinity related logic by default on supporting
kernels for cases like socket-lb and north-south LB (nodeport) case.

Additionally, rename the macro to better reflect the shared usage.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
service: Decouple affinity logic from session affinity config

This is a preparatory commit for the follow-up datapath changes.
The affinity maps will be reused in case of socket-lb
to store UDP client socket to backend mapping in order
to keep connections uninterrupted when service backends change.

The affinity related code will be executed when a service has affinity
enabled, or socket-lb (aka host reachable services) option is enabled
for kube-proxy replacement mode.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf: Fix backend selection for unconnected UDP service endpoint updates

In case of socket-lb for unconnected UDP, we select a service backend
based on the service backends count. So we may end up selecting a wrong
backend for existing UDP connections in case of service backend updates.

Fix:

When session affinity is disabled, we reuse the session affinity map
to store mapping for a UDP client to its selected backend using
socket cookie.
Why not enable session affinity behavior?
Session affinity is a stronger affinity in the sense that all
subsequent connections from a client will go to the same
selected backend. This can lead to uneven load-balancing.

Upgrade/Downgrade considerations:

Introduce a new flag in affinity map key using reserved bits to mark
an entry in order to indicate if its storing netns or socket cookie.
In case of downgrade, such entries won't be references as the flag
wouldn't exist.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on daemon/cmd/datapath.go:
```diff
@@ -403,7 +403,7 @@ func (d *Daemon) initMaps() error {
 		lxcmap.LXCMap.DeleteAll()
 	}
 
-	if option.Config.EnableSessionAffinity {
```
Should we remove the `option.Config.EnableSessionAffinity` altogether, also from the agent itself as a flag given we intend anyway to always build it in for KPR? (We'd then need a small KPR doc update and upgrade guide note as well.)

aditighag replied:
Yes, I've filed - https://github.com/cilium/cilium/issues/17584.

pchaigno replied:
The behavior change for `EnableSessionAffinity` (i.e., always true) is introduced in this PR, so shouldn't this PR also include the documentation changes? After all, this PR is effectively removing support for `enable-session-affinity=false`.

pchaigno commented on daemon/cmd/datapath.go:
```diff
@@ -403,7 +403,7 @@ func (d *Daemon) initMaps() error {
 		lxcmap.LXCMap.DeleteAll()
 	}
 
-	if option.Config.EnableSessionAffinity {
```
Nit: GitHub struggle to fit the commit title (` service: Decouple affinity related operations from session affinity config`) on a single line because it's longer than 70 characters. You can easily fix that by dropping `related` from the phrase without loosing meaning I think.

pchaigno commented on pkg/datapath/maps/map.go:
```diff
@@ -198,7 +198,7 @@ func (ms *MapSweeper) RemoveDisabledMaps() {
 	}
 
-	if !option.Config.EnableSessionAffinity {
+	if !option.Config.EnableNodePort || !option.Config.EnableHostReachableServices {
```
Would it make sense to create a helper function `BackendAffinityEnabled()` in pkg/config? You could then reuse it in all places modified in this commit. We already have e.g. `KubeProxyReplacementFullyEnabled()` for a similar purpose.

pchaigno commented on pkg/service/service.go:
```diff
@@ -730,33 +727,22 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
-			}
-		}
+	// When service affinity is disabled, the map is used to store socket affinity for UDP backends.
+	if prevSessionAffinity || option.Config.EnableHostReachableServices {
```
If `sessionAffinity` for that service was previously enabled but isn't anymore and `EnableHostReachableServices` is disabled, shouldn't we remove the entry from the affinity match map?

aditighag replied:
That's exactly what this check is doing.

pchaigno replied:
Yes, but it used to add the `svc.backends` to `toDeleteAffinity` and it doesn't anymore. If `option.Config.EnableHostReachableServices=false` and `svc.sessionAffinity=false`, shouldn't we still remove those `svc.backends`?

pchaigno commented on pkg/service/service_test.go:
```diff
@@ -710,3 +716,32 @@ func (m *ManagerTestSuite) TestLocalRedirectServiceOverride(c *C) {
+	c.Assert(len(m.lbmap.AffinityMatch[uint16(id1)]), Equals, 2)
+	for bID := range m.lbmap.BackendByID {
+		c.Assert(m.lbmap.AffinityMatch[uint16(id1)][bID], Equals, struct{}{})
+	}
```
Should this be extended to also cover changes of `svc.SessionAffinity` given you touched that logic in `upsertServiceIntoLBMaps`?

aditighag replied:
Unless I misunderstood your comment, there are already existing tests that test the relevant code around `svc.SessionAffinity`. As the title says, this test particularly tests the case where the config is disabled.

pchaigno commented on bpf/bpf_sock.c:
```diff
@@ -382,32 +382,19 @@ static __always_inline int __sock4_xlate_fwd(struct bpf_sock_addr *ctx,
-				 * doesn't hit the reselection again.
-				 */
-				backend_id = 0;
-		}
```
This commit (`bpf: Fix backend selection for unconnected UDP service endpoint updates`) would be a lot easier to review if the code refactoring (e.g., new `lb6_backend_id_by_affinity` function) was separated from the code changes.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -758,6 +758,91 @@ lb6_update_affinity_by_netns(const struct lb6_service *svc __maybe_unused,
+	memset(&key, 0, sizeof(key));
+	key.rev_nat_id   = svc->rev_nat_index;
+	key.sock_cookie  = 1;
+	key.client_id.client_sock_cookie    = id->client_sock_cookie;
```
Nit: Not sure what happened with the alignments here :-)

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -758,6 +758,91 @@ lb6_update_affinity_by_netns(const struct lb6_service *svc __maybe_unused,
+	if (val != NULL) {
+		struct lb_affinity_match match = {
+				.rev_nat_id	= svc->rev_nat_index,
+				.backend_id	= val->backend_id,
```
Nit: Same here regarding indentation.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -758,6 +758,91 @@ lb6_update_affinity_by_netns(const struct lb6_service *svc __maybe_unused,
+
+		return val->backend_id;
+	}
+#endif /* ENABLE_BACKEND_AFFINITY && BPF_HAVE_SOCKET_COOKIE */
```
We need to document this new "requirement" and what happens if kernel support is missing.

pchaigno commented on bpf/lib/lb.h:
```diff
@@ -758,6 +758,91 @@ lb6_update_affinity_by_netns(const struct lb6_service *svc __maybe_unused,
+lb6_affinity_backend_id_by_sock(const struct lb6_service *svc __maybe_unused,
+				const union lb6_affinity_client_id *id __maybe_unused)
+{
+#if defined(ENABLE_BACKEND_AFFINITY) && defined(BPF_HAVE_SOCKET_COOKIE)
```
Would it be worth probing `BPF_HAVE_SOCKET_COOKIE` from the agent and logging some warning/info if kernel support is missing? I'm not sure what kernel that corresponds to.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -1195,6 +1195,28 @@ func (kub *Kubectl) Logs(namespace string, pod string) *CmdRes {
 		fmt.Sprintf("%s -n %s logs %s", KubectlCmd, namespace, pod))
 }
 
```
Nit: This commit's title can be shortened to ` test: Validate UDP connectivity on service endpoints updates`.

pchaigno commented on test/helpers/kubectl.go:
```diff
@@ -1195,6 +1195,28 @@ func (kub *Kubectl) Logs(namespace string, pod string) *CmdRes {
 		fmt.Sprintf("%s -n %s logs %s", KubectlCmd, namespace, pod))
 }
 
+// LogsWithPrevious returns a CmdRes with containing the resulting metadata from the
```
```suggestion
// LogsWithPrevious returns a CmdRes with metadata from the
```
Same for the sibling functions below.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			clientPodLabel = "app=affinity-client"
+			serverPodLabel = "app=affinity-server"
+			serverRs = "affinity-server"
+			scaleUpReplicas = 3
```
These four variables can be made into constants.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			res = kubectl.ApplyDefault(affinityServerYAML)
+			Expect(res).Should(helpers.CMDSuccess(), "unable to apply %s", affinityServerYAML)
+			res = kubectl.ApplyDefault(affinityClientYAML)
+			Expect(res).Should(helpers.CMDSuccess(), "unable to apply %s", affinityClientYAML)
```
Could we merge these three manifests to simplify the code here and reduce the number of `kubectl` calls?

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			monitorCancel1()
+			monitorCancel2()
+			helpers.WriteToReportFile(monitorRes1.CombineOutput().Bytes(), "affinity-cilium-monitor-k8s1.log")
+			helpers.WriteToReportFile(monitorRes2.CombineOutput().Bytes(), "affinity-cilium-monitor-k8s2.log")
```
Is this really needed? We already have the Hubble logs for each node under `cilium-xxxxx-hubble_observe.txt` in the sysdump.

aditighag replied:
Ah! I didn't know that we already collect Hubble logs.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+
+			// Wait for all client and server pods to be in ready state.
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=affinity-server", helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```
Could we add an error message here? Most likely failure case will be timeout which usually doesn't provide a lot of useful information on its own. (One goal of error messages is to avoid having to resort to looking at the test code every time a test fails.)

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			Expect(err).Should(BeNil())
+			// Wait for the service backends to be plumbed in BPF maps
+			podIPs, err := kubectl.GetPodsIPs(helpers.DefaultNamespace, serverPodLabel)
+			ExpectWithOffset(1, err).Should(BeNil(), "Cannot retrieve pod IPs for %s", serverPodLabel)
```
Why `ExpectWithOffset(1, ...`?

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// Wait for the service backends to be plumbed in BPF maps
+			podIPs, err := kubectl.GetPodsIPs(helpers.DefaultNamespace, serverPodLabel)
+			ExpectWithOffset(1, err).Should(BeNil(), "Cannot retrieve pod IPs for %s", serverPodLabel)
+			ExpectWithOffset(1, len(podIPs)).To(Equal(2), "Expected pod IPs mismatch")
```
```suggestion
			ExpectWithOffset(1, len(podIPs)).To(Equal(2), "Unexpected number of service backends")
```

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+				err = kubectl.WaitForServiceBackend(helpers.K8s1, ip)
+				ExpectWithOffset(1, err).Should(BeNil(), "Failed waiting for %s backend entry on k8s1", ip)
+				err = kubectl.WaitForServiceBackend(helpers.K8s2, ip)
+				ExpectWithOffset(1, err).Should(BeNil(), "Failed waiting for %s backend entry on k8s2", ip)
```
Could this be replaced by a call to `ValidateServicePlumbing`?

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+				"-l app=affinity-client").Filter(filter)
+			ExpectWithOffset(1, err).To(BeNil(), "Failed to query affinity-client pod")
+			restartCount, err := strconv.Atoi(countRes.String())
+			ExpectWithOffset(1, err).To(BeNil(), "Failed to convert count value")
```
This code could be moved to a helper function given it's needed again at the end of the test.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			res := kubectl.Exec(cmd)
+			res.ExpectSuccess(fmt.Sprintf("Failed to scale up replica set: %s", serverRs))
+			err = kubectl.WaitforPods(helpers.DefaultNamespace, "-l zgroup=affinity-server", helpers.HelperTimeout)
+			Expect(err).Should(BeNil())
```
Could we add an error message here?

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// that client connectivity was interrupted, and that subsequent messages
+			// from the client were sent to a different service endpoint.
+			res = kubectl.Logs(helpers.DefaultNamespace, clientPod)
+			res.ExpectDoesNotContain("server reply mismatch", "connectivity interrupted on client pod %s", clientPod)
```
How do we know that, at this point, the client attempted to connect to the server again?

aditighag replied:
I've fixed the test - the problem was that client was making a `DialUDP` call in a loop until it can connect to the server for the very first time. But based on the documentation, the function, when passed a hostname, only resolves the hostname. It won't actually connect to the server endpoint as I originally thought it would. I've updated the test code to match on the initial server reply instead.

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// }
+		})
+
+		It("Checks clusterIP connectivity is uninterrupted on service endpoints' updates", func() {
```
Our code to generate Jenkins artifacts often struggles with special characters. We could spend more time fixing it, but TBH it's easier to just rename the test :-)
```suggestion
		It("Checks clusterIP connectivity is uninterrupted on service endpoint updates", func() {
```

pchaigno commented on test/k8sT/Services.go:
```diff
@@ -1547,6 +1548,143 @@ Secondary Interface %s :: IPv4: (%s, %s), IPv6: (%s, %s)`, helpers.DualStackSupp
+			// }
+		})
+
+		It("Checks clusterIP connectivity is uninterrupted on service endpoints' updates", func() {
```
Adding a couple `By(...)` in the test would help parsing the test output afterward. See other tests for examples.

For example, here we could have:
1. `By("Counting number of restarts of client pod")`
2. `By("Scaling up the server replica set")`
3. `By("Waiting for backend map to be up-to-date")`
4. `By("Checking client logs for interrupted connections")`
5. `By("Checking client pod didn't restart")`

It has the nice side-effect of documenting (a little) the tests :-)

aditighag replied:
Nice suggestion!

pchaigno reviewed (changes requested):
In addition to inline comments below, documentation is missing and checkpatch has several legitimate complains.

I had a quick look at the failing test but couldn't figure it out :-(

aditighag committed:
config/docs: Requirements for backend affinity implementation

Adds kernel support checks/warnings, and document the requirements.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1187,7 +1187,6 @@ replacement. For users who run on older kernels which do not support the network
 namespace cookies, a fallback in-cluster mode is implemented, which is based on
 a fixed cookie value as a trade-off. This makes all applications on the host to
 select the same service endpoint for a given service with session affinity configured.
-To disable the feature, set ``config.sessionAffinity=false``.
```
I've extended the document with the new behavior - 

```
The configuration option ``config.sessionAffinity`` will be deprecated
in future, and doesn't disable the feature. In order to disable session affinity
for a service, remove the Kubernetes session affinity setting for the service.
```

borkmann commented on daemon/cmd/kube_proxy_replacement.go:
```diff
@@ -391,6 +364,39 @@ func initKubeProxyReplacementOptions() (bool, error) {
+
+		if h := probesManager.GetHelpers("cgroup_sock"); h != nil {
+			if _, ok := h["bpf_get_socket_cookie"]; !ok && option.Config.EnableHostReachableServices {
+				log.Warn("UDP connections may experience disruptions on " +
```
Do you happen to know for which kernel this would be the case? Given we didn't warn about this before, I wouldn't specifically add a new one here since this might confuse users as it needs more explanation wrt what's going on and if this is an issue for their env.

aditighag replied:
>Do you happen to know for which kernel this would be the case? 

Well, if the relevant config option isn't set. We can drop this warning since the requirement is documented in the getting started guide.

borkmann commented on Documentation/gettingstarted/kubeproxy-free.rst:
```diff
@@ -1197,6 +1197,13 @@ to the same service endpoints; but two requests for the same service, sent from
+To ensure that (unconnected) UDP client socket level connections to service
+endpoints are not interrupted when endpoints are updated, run Linux kernel > 5.3.
+Specifically, the datapath implementation requires BPF LRU (*Least Recently Used*)
+maps and ``bpf_get_socket_cookie`` helper that returns a globally unique identifier.
```
Could we merge this into the previous 'session affinity' paragraph, stating that for unconnected UDP services the agent implements an implicit affinity for kernels 5.4 and beyond. (I don't think we need to outline the implementation details for users in here.)

borkmann commented on pkg/service/service.go:
```diff
@@ -730,32 +730,33 @@ func (s *Service) upsertServiceIntoLBMaps(svc *svcInfo, onlyLocalBackends bool,
+			if !option.Config.EnableHostReachableServices {
+				toDeleteAffinity = make([]lb.BackendID, 0, len(svc.backends))
+				for _, b := range svc.backends {
+					toDeleteAffinity = append(toDeleteAffinity, b.ID)
```
Cc @brb for more in-depth review, but I think there's a bug in here: if we hit this situation, then we will delete the current svc backends, but not the obsolete ones given we override the scope of toDeleteAffinity in above branch. (Old code had `make([]lb.BackendID, 0, len(obsoleteSVCBackendIDs)+len(svc.backends))`.)

aditighag replied:
We need to delete obsolete svc backends as long as the previous session affinity was enabled, the current service affinity case doesn't matter.  The [previous code](https://github.com/cilium/cilium/blob/master/pkg/service/service.go#L749) did have this logic, but it was written in a roundabout way. I simplified the logic - 
we delete the obsolete ones in L743-746. The unit test cases validate the different scenarios.

borkmann replied:
> we delete the obsolete ones in L743-746

Right, but wouldn't `toDeleteAffinity = make([]lb.BackendID, 0, len(svc.backends))` override this again?

aditighag replied:
Ah! You are right. Thanks! I'll add a new unit test case.

aditighag replied:
I found one more bug while adding new unit test cases. The logic should be robust now.

borkmann commented on bpf/lib/lb.h:
```diff
@@ -758,6 +758,91 @@ lb6_update_affinity_by_netns(const struct lb6_service *svc __maybe_unused,
+		 * for the same service:port. lb6_update_affinity_by_netns()
+		 * below would then override the first created one if it
+		 * didn't make it into the lookup yet for the other CPU.
+		 */
```
That would also hold true for the socket cookie on unconnected udp, no?

borkmann commented on bpf/lib/lb.h:
```diff
@@ -1309,6 +1394,90 @@ lb4_update_affinity_by_netns(const struct lb4_service *svc __maybe_unused,
+		}
+
+		return val->backend_id;
+	}
```
nit: could we reuse most of __lb{4,6}_affinity_backend_id?
